cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,528 @@
+
+```protobuf
+message NodeGroupForNodeRequest {
+  // Node group for the given node
+  ExternalGrpcNode node = 1;
+}
+
+message NodeGroupForNodeResponse {
+  // The node group for the given node.
+  repeated NodeGroup nodeGroup = 1;
+}
+
+message NodeGroupForNodeRequest {
+  // Node group for the given node
+  ExternalGrpcNode node = 1;
+}
+
+```
+
+### GetAvailableGPUTypes
+
+GetAvailableGPUTypes handles all available GPU types cloud provider supports.
+
+```protobuf
+message GetAvailableGPUTypesRequest {
+  // Intentionally empty.
+}
+
+message GetAvailableGPUTypesResponse {
+ // GPU types passed in as opaque key-value pairs.
+  map<string, Any> gpuTypes = 1;
+}
+```
+
+### GPULabel
+
+GPULabel stores the label added to nodes with GPU resource.
+
+```protobuf
+message GPULabelRequest {
+  // Intentionally empty.
+}
+
+message GPULabelResponse {
+  // Label value of the GPU
+ string label = 1;
+}
+```
+
+### PricingNodePrice
+
+NodePrice handles an operation that returns a price of running the given node for a given period of time.
+PricingNodePrice is an optional operation that is not implemented by all the providers.
+
+```protobuf
+message PricingNodePriceRequest {
+  ExternalGrpcNode node = 1;,
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time startTime = 2;
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time endTime = 3;
+}
+
+message PricingNodePriceResponse {
+ // Price of the theoretical minimum price of running a pod for a given period
+ float64 price = 1;
+}
+```
+
+### PricingPodPrice
+
+PodPrice handles an operation that returns a theoretical minimum price of running a pod for a given period of time on a perfectly matching machine.
+PricingPodPrice is an optional operation that is not implemented by all the providers.
+
+
+```protobuf
+message PricingPodPriceRequest {
+  k8s.io.api.core.v1.Pod pod = 1;,
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time startTime = 2;
+
+  k8s.io.apimachinery.pkg.apis.meta.v1.Time endTime = 3;
+}
+
+message PricingPodPriceResponse {
+ // Price of the theoretical minimum price of running a pod for a given period
+ float64 price = 1;
+}
+```
+
+### Refresh
+
+Refresh is called before every main loop and can be used to dynamically update cloud provider state.
+In particular the list of node groups returned by NodeGroups can change as a result of this action.
+
+```protobuf
+message RefreshRequest {
+  // Intentionally empty.
+}
+
+message RefreshResponse {
+// Intentionally empty.
+}
+```
+
+### Cleanup
+
+Cleanup cleans up open resources before the cloud provider is destroyed, i.e. go routines etc.
+
+```protobuf
+message CleanupRequest {
+  // Intentionally empty.
+}
+
+message CleanupResponse {
+// Intentionally empty.
+}
+```
+
+### NodeGroup
+
+```protobuf
+message NodeGroup {
+  // ID of the node group on the cloud provider
+  string id = 1;
+
+  // MinSize of the node group on the cloud provider
+  int32 minSize = 2;
+
+  // MaxSize of the node group on the cloud provider
+  int32 maxSize = 3;
```

Comment:
I'm not sure if we should actually include min, max and debug. min/max can change (after Refresh() call), while id is the consistent group identifier. For something like NodeGroupForNode it seems more consistent to return permanent id and not some settings that can change later on.